dali/pipeline/operators/reader/reader_op.h

```diff
@@ -47,217 +47,191 @@ namespace dali {
-        }
-
-        // signal the other workers we're ready
-        prefetch_ready_workers_ = true;
-
-        // signal the prefetch thread to start again
-        producer_.notify_one();
-      }
-    }
+    // If necessary start prefetching thread and wait for a consumable batch
+    StartPrefetchThread();
+    ConsumerWait();
 
-    // consume batch
+    // consume sample
+    TimeRange tr("DataReader::Run #" + to_string(curr_batch_consumer_), TimeRange::kViolet);
     Operator<Backend>::Run(ws);
-
-    loader_->ReturnTensor(prefetched_batch_[ws->data_idx()]);
-    prefetched_batch_[ws->data_idx()] = nullptr;
-
+    auto *sample = GetSample(ws->data_idx());
+    loader_->ReturnTensor(sample);
+    sample = nullptr;
     samples_processed_++;
 
-    // lock, check if batch is finished, notify
-    {
-      std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
-
-      // if we need to stop, stop.
-      if (batch_stop_) return;
-
-      // if we've consumed all samples in this batch, reset state and stop
-      if (samples_processed_.load() == Operator<Backend>::batch_size_) {
-        prefetch_ready_workers_ = false;
-        prefetched_batch_ready_ = false;
-        producer_.notify_one();
-        samples_processed_ = 0;
-        batch_stop_ = true;
-      }
-      return;
+    // if we've processed the whole batch, notify it
+    if (samples_processed_.load() == Operator<Backend>::batch_size_) {
+      samples_processed_ = 0;
+      ConsumerAdvanceQueue();
     }
   }
 
   // GPUBackend operators
   void Run(DeviceWorkspace* ws) override {
+    // If necessary start prefetching thread and wait for a consumable batch
     StartPrefetchThread();
+    ConsumerWait();
 
-    // grab the actual prefetching lock
-    std::unique_lock<std::mutex> prefetch_lock(prefetch_access_mutex_);
+    // Consume batch
+    Operator<Backend>::Run(ws);
+    CUDA_CALL(cudaStreamSynchronize(ws->stream()));
+    for (auto &sample : prefetched_batch_queue_[curr_batch_consumer_]) {
+      loader_->ReturnTensor(sample);
+    }
 
-    // Wait until prefetch is ready
-    consumer_.wait(prefetch_lock, [this]() { return prefetched_batch_ready_; });
+    // Notify we have consumed a batch
+    ConsumerAdvanceQueue();
+  }
 
-    if (prefetch_error_) {
-      DALI_FAIL("Prefetching failed: " + dali::string(*prefetch_error_));
-    }
+  Index epoch_size() const override {
+    return loader_->Size();
+  }
 
-    producer_.notify_one();
+  LoadTarget* GetSample(int sample_idx) {
+    return prefetched_batch_queue_[curr_batch_consumer_][sample_idx];
+  }
 
-    Operator<Backend>::Run(ws);
+ protected:
+  void ProducerStop(std::exception_ptr error = nullptr) {
+    std::lock_guard<std::mutex> lock(prefetch_access_mutex_);
+    finished_ = true;
+    if (error) prefetch_error_ = error;
+    consumer_.notify_all();
+  }
 
-    CUDA_CALL(cudaStreamSynchronize(ws->stream()));
+  void ProducerAdvanceQueue() {
+    std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
```

Comment:
You can use lock_guard here, and also unlock it before you notify.

Reply:
will do